### ISSUE TITLE
voc2012test

### ISSUE THREAD

**Issue by foralliance** [weiliu89/caffe#848]
how to test voc2012??
the code "score_ssd_pascal.py" Seems to fit only voc2007. because for voc2012, there is no VOC0712Plus_test_lmdb.

**Comment by foralliance** [comment]
Purpose:
I want to use the " VGG_VOC0712Plus_SSD_300x300_iter_240000. Caffemodel" to test the voc2012test dataset.

Method1: 
with the help of the test file in [Refinedet](https://github.com/sfzhang15/RefineDet) , the result of voc2012test was obtained. After submitting the server, the result was 0.75960. It's OK

Method2:
because author didn't directly provide the test code for voc2012test, I imitate " score_ssd_pascal.py " write a code about voc2012test ,called " score_ssd_pascal0712plus.py ". In particular, it is simple to modify some training set names, file path names, and test the number of images.

because the "create_data. sh" ( from data/VOC0712Plus) gets only LMDB files on the training set(VOC0712Plus_trainval_lmdb), in order to get LMDB files about test set, imitate "create_data. sh" (from data/VOC0712) to write,just change
   for subset in trainval 
to
   for subset in test trainval

then, running, point out:
Annofation file: / home/JMX/data/VOCdevkit VOC2012 / Annotations / 2008 _000001. XML does not exist

so I randomly wrote a " 2008_000001.xml file " under the Annotations folder, and run successfully,  get " VOC0712Plus_test_lmdb "

then run "score_ssd_pascal0712plus.py" and get the result to the server, but the final result is only 0.32190.

----------------------------------------------------------------------------------------------------------------------------------

Why is the result of method 2 so bad? What went wrong?

**Comment by foralliance** [comment]
Purpose:
I want to use the " VGG_VOC0712Plus_SSD_300x300_iter_240000. Caffemodel" to test the voc2012test dataset.

Method1:
with the help of the test file in Refinedet , the result of voc2012test was obtained. After submitting the server, the result was 0.75960. It's OK

Method2:
because author didn't directly provide the test code for voc2012test, I imitate " score_ssd_pascal.py " write a code about voc2012test ,called " score_ssd_pascal0712plus.py ". In particular, it is simple to modify some training set names, file path names, and test the number of images.

because the "create_data. sh" ( from data/VOC0712Plus) gets only LMDB files on the training set(VOC0712Plus_trainval_lmdb), in order to get LMDB files about test set, imitate "create_data. sh" (from data/VOC0712) to write,just change
for subset in trainval
to
for subset in test trainval

then, running, point out:
**Annofation file: / home/JMX/data/VOCdevkit VOC2012 / Annotations / 2008 _000001. XML does not exist**

so I randomly wrote a " 2008_000001.xml file " under the Annotations folder, and run successfully, get " VOC0712Plus_test_lmdb "

then run "score_ssd_pascal0712plus.py" and get the result to the server, but the final result is only 0.32190.

Why is the result of method 2 so bad? What went wrong?

**Comment by foralliance** [comment]
Purpose:
I want to use the " VGG_VOC0712Plus_SSD_300x300_iter_240000. Caffemodel" to test the voc2012test dataset.

Method1:
with the help of the test file in Refinedet , the result of voc2012test was obtained. After submitting the server, the result was 0.75960. It's OK

Method2:
because author didn't directly provide the test code for voc2012test, I imitate " score_ssd_pascal.py " write a code about voc2012test ,called " score_ssd_pascal0712plus.py ". In particular, it is simple to modify some training set names, file path names, and test the number of images.

because the "create_data. sh" ( from data/VOC0712Plus) gets only LMDB files on the training set(VOC0712Plus_trainval_lmdb), in order to get LMDB files about test set, imitate "create_data. sh" (from data/VOC0712) to write,just change
for subset in trainval
to
for subset in test trainval

then, running, point out:
`Annofation file: / home/JMX/data/VOCdevkit VOC2012 / Annotations / 2008 _000001. XML does not exist`

so I randomly wrote a " 2008_000001.xml file " under the Annotations folder, and run successfully, get " VOC0712Plus_test_lmdb "

then run "score_ssd_pascal0712plus.py" and get the result to the server, but the final result is only 0.32190.

Why is the result of method 2 so bad? What went wrong?

**Comment by foralliance** [comment]
Purpose:
I want to use the " VGG_VOC0712Plus_SSD_300x300_iter_240000. Caffemodel" to test the voc2012test dataset.

Method1:
with the help of the test file in [Refinedet](https://github.com/sfzhang15/RefineDet) , the result of voc2012test was obtained. After submitting the server, the result was 0.75960. It's OK

Method2:
because author didn't directly provide the test code for voc2012test, I imitate " score_ssd_pascal.py " write a code about voc2012test ,called " score_ssd_pascal0712plus.py ". In particular, it is simple to modify some training set names, file path names, and test the number of images.

because the "create_data. sh" ( from data/VOC0712Plus) gets only LMDB files on the training set(VOC0712Plus_trainval_lmdb), in order to get LMDB files about test set, imitate "create_data. sh" (from data/VOC0712) to write,just change
`for subset in trainval`
to
`for subset in test trainval`

then, running, point out:
`Annofation file: / home/JMX/data/VOCdevkit VOC2012 / Annotations / 2008 _000001. XML does not exist`

so I randomly wrote a " 2008_000001.xml file " under the Annotations folder, and run successfully, get " VOC0712Plus_test_lmdb "

then run "score_ssd_pascal0712plus.py" and get the result to the server, but the final result is only 0.32190.

----------------------------------------------------------------------------------------------------------------------------------

Why is the result of method 2 so bad? What went wrong?

**Comment by foralliance** [comment]
Purpose:
I want to use the " VGG_VOC0712Plus_SSD_300x300_iter_240000. Caffemodel" to test the voc2012test dataset.

Method1:
with the help of the test file in [Refinedet](https://github.com/sfzhang15/RefineDet) , the result of voc2012test was obtained. After submitting the server, the result was 0.75960. It's OK

Method2:
because author didn't directly provide the test code for voc2012test, I imitate "`score_ssd_pascal.py`" write a code about voc2012test ,called "`score_ssd_pascal0712plus.py`". In particular, it is simple to modify some training set names, file path names, and test the number of images.

because the "`create_data. sh`" ( from data/VOC0712Plus) gets only LMDB files on the training set(VOC0712Plus_trainval_lmdb), in order to get LMDB files about test set, imitate "`create_data. sh`" (from data/VOC0712) to write,just change
`for subset in trainval`
to
`for subset in test trainval`

then, running, point out:
`Annofation file: /home/xxx/data/VOCdevkit/VOC2012/Annotations/2008_000001.xml does not exist`

so I randomly wrote a "2008_000001.xml file " under the Annotations folder, and run successfully, get " VOC0712Plus_test_lmdb "

then run "score_ssd_pascal0712plus.py" and get the result to the server, but the final result is only 0.32190.

----------------------------------------------------------------------------------------------------------------------------------

Why is the result of method 2 so bad? What went wrong?

**Comment by foralliance** [comment]
Purpose:
I want to use the " **VGG_VOC0712Plus_SSD_300x300_iter_240000. Caffemodel**" to test the **voc2012test dataset**.

Method1:
with the help of the test file in [Refinedet](https://github.com/sfzhang15/RefineDet) , the result of voc2012test was obtained. After submitting the server, the result was **0.75960**. It's OK

Method2:
because author didn't directly provide the test code for voc2012test, I imitate "`score_ssd_pascal.py`" write a code about voc2012test ,called "`score_ssd_pascal0712plus.py`". In particular, it is simple to modify some training set names, file path names, and test the number of images.

because the "`create_data. sh`" ( from data/VOC0712Plus) gets only LMDB files on the training set(**VOC0712Plus_trainval_lmdb**), in order to get LMDB files about test set, imitate "`create_data. sh`" (from data/VOC0712) to write,just change
`for subset in trainval`
to
`for subset in test trainval`

then, running, point out:
`Annofation file: /home/xxx/data/VOCdevkit/VOC2012/Annotations/2008_000001.xml does not exist`

so I randomly wrote a "2008_000001.xml file " under the Annotations folder, and run successfully, get "**VOC0712Plus_test_lmdb**"

then run "`score_ssd_pascal0712plus.py`" and get the result to the server, but the final result is only **0.32190**.

----------------------------------------------------------------------------------------------------------------------------------

Why is the result of method 2 so bad? What went wrong?

**Comment by shiyuangogogo** [comment]
@foralliance I meet the same problem. have you solve it?  I use ImageDataLayer to test voc12-test dataset.
but i got a bad result .

**Comment by moothes** [comment]
Maybe you have shuffled the test set？

**Comment by shiyuangogogo** [comment]
@moothes each score is around 0.01-0.04, it's bad. I hadn't shuffle the test set. And i think whether shuffle doesn't matter.

**Comment by shiyuangogogo** [comment]
I use ImageDataLayer to test voc12-test,  mAP=0.75640

**Comment by foralliance** [comment]
Hi, @moliqian 
what's your mean "ImageDataLayer to test"? specifically, what is your code?

**Comment by shiyuangogogo** [comment]
ImageDataLayer is one layer of Caffe to load image to CNN

**Comment by frostinassiky** [comment]
@foralliance Better than me. I also used fake annotations but got 0.298 mAP online. Are there any good ways to test VOC2012?

**Comment by XuanheLiu** [comment]
Where can you download the correct voc2012 test?

**Comment by frostinassiky** [comment]
@XuanheLiu This is not possible

**Comment by XuanheLiu** [comment]
@Frostinassiky Do everyone do experiments to take voc2012test to test, do they have to annotate data sets themselves?

**Comment by ryusaeba** [comment]
Hi @moliqian  ,
I thought ImageData Layer is for classification task. How do you use ImageDataLayer as SSD's input? 
Another question is do you implement data augmentation in the layer? 
Could you please share the code with me? thanks!

**Comment by Zxl19990529** [comment]
I have also met such a problem. I download VOCtest_2012 here:[https://pjreddie.com/projects/pascal-voc-dataset-mirror/]( https://pjreddie.com/projects/pascal-voc-dataset-mirror/) .
But there's no corresponding annotation files recorded in ImageSets/Main/test.txt .

**Comment by moothes** [comment]
@Zxl19990529 
To test models on VOC2012 set, you can:
1、Generate some fake annotation files, for example, annotating each image with same boxes; 
2、Run SSD testing code to detect on VOC2012 test set. You will get a very low mAP and it doesn't matter;
3、Copy the detection result files and upload them to official website.
I get the VOC2012 test result through above steps. Here is our [VOC12 result](http://host.robots.ox.ac.uk:8080/anonymous/CYWQES.html) .
The model is our ICCV2017 workshop paper [CAD net](https://ieeexplore.ieee.org/document/8265304) . 
Sorry for my poor english.

**Comment by Zxl19990529** [comment]
@moothes  Aha, I see, thank you very much.

**Comment by keyang1** [comment]
@Frostinassiky I also get the problem,my result is 0.32,do you solve the problem?

**Comment by keyang1** [comment]
@moothes I do the all step as you ,but I get the result is 0.32,could you help me to solve the problem?
